File: src/main/java/com/github/sorusclient/client/transform/Applier.java
package com.github.sorusclient.client.transform;

import org.objectweb.asm.tree.*;

import java.util.function.Consumer;
import java.util.function.Function;

public class Applier {

    public static class Insert<T> implements Consumer<MethodNode> {

        private final T instruction;
        private final Function<MethodNode, T> function;

        public Insert(T instruction) {
            this.instruction = instruction;
            this.function = null;
        }

        public Insert(Function<MethodNode, T> function) {
            this.function = function;
            this.instruction = null;
        }

        @Override
        public void accept(MethodNode methodNode) {
            T instructionT = this.instruction != null ? this.instruction : this.function.apply(methodNode);
            Object instruction = Applier.clone(instructionT);
            if (instruction instanceof InsnList) {
                methodNode.instructions.insert((InsnList) instruction);
            } else if (this.instruction instanceof AbstractInsnNode) {
                methodNode.instructions.insert((AbstractInsnNode) instruction);
            }
        }

    }

    public static class InsertBefore<T, U extends AbstractInsnNode> implements Consumer<U> {

        private final MethodNode methodNode;
        private final T instruction;

        public InsertBefore(MethodNode methodNode, T instruction) {
            this.methodNode = methodNode;
            this.instruction = instruction;
        }

        @Override
        public void accept(U node) {
            Object instruction = Applier.clone(this.instruction);
            if (instruction instanceof InsnList) {
                this.methodNode.instructions.insertBefore(node, (InsnList) instruction);
            } else if (this.instruction instanceof AbstractInsnNode) {
                this.methodNode.instructions.insertBefore(node, (AbstractInsnNode) instruction);
            }
        }

    }

    public static class InsertAfter<T, U extends AbstractInsnNode> implements Consumer<U> {

        private final MethodNode methodNode;
        private final T instruction;

        public InsertAfter(MethodNode methodNode, T instruction) {
            this.methodNode = methodNode;
            this.instruction = instruction;
        }

        @Override
        public void accept(U node) {
            Object instruction = Applier.clone(this.instruction);
            if (instruction instanceof InsnList) {
                this.methodNode.instructions.insert(node, (InsnList) instruction);
            } else if (this.instruction instanceof AbstractInsnNode) {
                this.methodNode.instructions.insert(node, (AbstractInsnNode) instruction);
            }
        }

    }

    private static Object clone(Object instruction) {
        if (instruction instanceof InsnList) {
            InsnList insnList = new InsnList();
            for (AbstractInsnNode node : ((InsnList) instruction)) {
                insnList.add((AbstractInsnNode) Applier.clone(node));
            }
            return insnList;
        } else if (instruction instanceof MethodInsnNode) {
            MethodInsnNode methodInsnNode = (MethodInsnNode) instruction;
            return new MethodInsnNode(methodInsnNode.getOpcode(), methodInsnNode.owner, methodInsnNode.name, methodInsnNode.desc);
        } else if (instruction instanceof VarInsnNode) {
            VarInsnNode varInsnNode = (VarInsnNode) instruction;
            return new VarInsnNode(varInsnNode.getOpcode(), varInsnNode.var);
        } else if (instruction instanceof InsnNode) {
            InsnNode insnNode = (InsnNode) instruction;
            return new InsnNode(insnNode.getOpcode());
        } else if (instruction instanceof JumpInsnNode) {
            JumpInsnNode jumpInsnNode = (JumpInsnNode) instruction;
            return new JumpInsnNode(jumpInsnNode.getOpcode(), (LabelNode) Applier.clone(jumpInsnNode.label));
        } else if (instruction instanceof LabelNode) {
            LabelNode labelNode = (LabelNode) instruction;
            return new LabelNode(labelNode.getLabel());
        }
        return null;
    }

}

File: src/main/kotlin/com/github/sorusclient/client/transform/Applier.kt
/*
 * Copyright (c) 2022. Danterus
 * Copyright (c) 2022. Sorus Contributors
 *
 * SPDX-License-Identifier: MPL-2.0
 */

package com.github.sorusclient.client.transform

import org.objectweb.asm.tree.*
import java.util.function.Consumer
import java.util.function.Function

object Applier {

    private fun clone(instruction: Any): Any? {
        when (instruction) {
            is InsnList -> {
                val insnList = InsnList()
                for (node in instruction) {
                    insnList.add(clone(node) as AbstractInsnNode?)
                }
                return insnList
            }
            is MethodInsnNode -> {
                return MethodInsnNode(
                    instruction.opcode,
                    instruction.owner,
                    instruction.name,
                    instruction.desc,
                    instruction.itf
                )
            }
            is FieldInsnNode -> {
                return FieldInsnNode(
                    instruction.opcode,
                    instruction.owner,
                    instruction.name,
                    instruction.desc
                )
            }
            is VarInsnNode -> {
                return VarInsnNode(instruction.opcode, instruction.`var`)
            }
            is InsnNode -> {
                return InsnNode(instruction.opcode)
            }
            is JumpInsnNode -> {
                return JumpInsnNode(instruction.opcode, clone(instruction.label) as LabelNode?)
            }
            is LabelNode -> {
                return LabelNode(instruction.label)
            }
            is TypeInsnNode -> {
                return TypeInsnNode(instruction.opcode, instruction.desc)
            }
            is LdcInsnNode -> {
                return LdcInsnNode(instruction.cst)
            }
            else -> return null
        }
    }

    class Insert : Consumer<MethodNode> {
        private val instruction: Any?
        private val function: Function<MethodNode, Any>?

        constructor(instruction: InsnList) {
            this.instruction = instruction
            function = null
        }

        constructor(instruction: AbstractInsnNode) {
            this.instruction = instruction
            function = null
        }

        constructor(function: Function<MethodNode, Any>?) {
            this.function = function
            instruction = null
        }

        override fun accept(methodNode: MethodNode) {
            val instructionT = instruction ?: function!!.apply(methodNode)
            val instruction = clone(instructionT)
            if (instruction is InsnList) {
                methodNode.instructions.insert(instruction)
            } else if (this.instruction is AbstractInsnNode) {
                methodNode.instructions.insert(instruction as AbstractInsnNode)
            }
        }
    }

    class InsertBefore<T: Any, U : AbstractInsnNode?>(private val methodNode: MethodNode, private val instruction: T) :
        Consumer<U> {
        override fun accept(node: U) {
            val instruction = clone(instruction)
            if (instruction is InsnList) {
                methodNode.instructions.insertBefore(node, instruction as InsnList?)
            } else if (this.instruction is AbstractInsnNode) {
                methodNode.instructions.insertBefore(node, instruction as AbstractInsnNode?)
            }
        }
    }

    class InsertAfter<T: Any, U : AbstractInsnNode?>(private val methodNode: MethodNode, private val instruction: T) :
        Consumer<U> {
        override fun accept(node: U) {
            val instruction = clone(instruction)
            if (instruction is InsnList) {
                methodNode.instructions.insert(node, instruction as InsnList?)
            } else if (this.instruction is AbstractInsnNode) {
                methodNode.instructions.insert(node, instruction as AbstractInsnNode?)
            }
        }
    }

}